test/src/IO/ReaderTest.cpp

```diff
@@ -75,45 +101,85 @@ namespace TrenchBroom {
             r.seekFromBegin(2U);
             EXPECT_EQ(2U, r.position());
 
-            EXPECT_THROW(r.seekFromBegin(3U), ReaderException);
+            EXPECT_THROW(r.seekFromBegin(11U), ReaderException);
             EXPECT_EQ(2U, r.position());
         }
 
-        TEST(ReaderTest, testSeekFromEnd) {
-            const char* foo = "xy";
-            auto r = Reader::from(foo, foo + 2);
+        TEST(BufferReaderTest, testSeekFromBegin) {
+            seekFromBegin(Reader::from(buff(), buff() + 10));
 
-            EXPECT_EQ(2U, r.size());
-            EXPECT_EQ(0U, r.position());
+        }
 
+        TEST(FileReaderTest, testSeekFromBegin) {
+            seekFromBegin(file()->reader());
+        }
+
+        void seekFromEnd(Reader&& r) {
             r.seekFromEnd(0U);
-            EXPECT_EQ(2U, r.position());
+            EXPECT_EQ(10U, r.position());
 
             r.seekFromEnd(1U);
-            EXPECT_EQ(1U, r.position());
+            EXPECT_EQ(9U, r.position());
 
-            r.seekFromEnd(2U);
+            r.seekFromEnd(10U);
             EXPECT_EQ(0U, r.position());
 
-            EXPECT_THROW(r.seekFromEnd(3U), ReaderException);
+            EXPECT_THROW(r.seekFromEnd(11U), ReaderException);
             EXPECT_EQ(0U, r.position());
         }
 
-        TEST(ReaderTest, testSeekForward) {
-            const char* foo = "xy";
-            auto r = Reader::from(foo, foo + 2);
+        TEST(BufferReaderTest, testSeekFromEnd) {
+            seekFromEnd(Reader::from(buff(), buff() + 10));
+        }
 
-            EXPECT_EQ(2U, r.size());
-            EXPECT_EQ(0U, r.position());
+        TEST(FileReaderTest, testSeekFromEnd) {
+            seekFromEnd(file()->reader());
+        }
 
+        void seekForward(Reader&& r) {
             r.seekForward(1U);
             EXPECT_EQ(1U, r.position());
 
             r.seekForward(1U);
             EXPECT_EQ(2U, r.position());
 
-            EXPECT_THROW(r.seekForward(1U), ReaderException);
+            EXPECT_THROW(r.seekForward(9U), ReaderException);
             EXPECT_EQ(2U, r.position());
         }
+
+        TEST(BufferReaderTest, testSeekForward) {
+            seekForward(Reader::from(buff(), buff() + 10));
+        }
+
+        TEST(FileReaderTest, testSeekForward) {
+            seekForward(file()->reader());
+        }
+
+        void subReader(Reader&& r) {
+            auto s = r.subReaderFromBegin(5, 3);
+
+            EXPECT_EQ(3U, s.size());
+            EXPECT_EQ(0U, s.position());
+
+            s.seekForward(1U);
```

Comment:
It'd be worth checking that we're actually reading from the correct place in the original array, i.e. replace this bit with:
```
EXPECT_EQ('f', s.readChar<char>());
EXPECT_EQ(1U, s.position());

EXPECT_EQ('g', s.readChar<char>());
EXPECT_EQ(2U, s.position());

EXPECT_EQ('h', s.readChar<char>());
EXPECT_EQ(3U, s.position());
```

Reply:
yup, done